source/adapters/cuda/usm.cpp

```diff
@@ -141,16 +141,21 @@ ur_result_t USMDeviceAllocImpl(void **ResultPtr, ur_context_handle_t,
     return Err;
   }
 
+  const bool validAlignment =
+      (Alignment == 0 ||
+       reinterpret_cast<std::uintptr_t>(*ResultPtr) % Alignment == 0);
 #ifdef NDEBUG
-  std::ignore = Alignment;
+  if (!validAlignment) {
+    urUSMFree(hContext, *ResultPtr);
+    return UR_RESULT_ERROR_UNSUPPORTED_ALIGNMENT;
```

Comment:
Given the old use of asserts are not checking an invariant I think this is a good direction. I do think we should be replacing the `#ifdef NDEBUG` logic with an always on clean up/error code path, though.

Currently `UR_RESULT_ERROR_UNSUPPORTED_ALIGNMENT` is not listed as a valid return code in [urUSMHostAlloc](https://oneapi-src.github.io/unified-runtime/core/api.html#urusmhostalloc)/[urUSMDeviceAlloc](https://oneapi-src.github.io/unified-runtime/core/api.html#urusmdevicealloc)/[urUSMSharedAlloc](https://oneapi-src.github.io/unified-runtime/core/api.html#_CPPv416urUSMSharedAlloc19ur_context_handle_t18ur_device_handle_tPK13ur_usm_desc_t20ur_usm_pool_handle_t6size_tPPv). Alignment check failures are currently returning `UR_RESULT_ERROR_INVALID_VALUE` instead. Either the spec should be updated to have `UR_RESULT_ERROR_UNSUPPORTED_ALIGNMENT` or the error code should align with the spec. 

Would need to give more thought which option is best, I think.

Reply:
Great catch, thanks!
I'll remove the condition execution from CUDA and will also replace UR_RESULT_ERROR_UNSUPPORTED_ALIGNMENT with UR_RESULT_ERROR_INVALID_VALUE, at least for now. It feels misleading to return an unsupported alignment error when the allocation call returned a non-null pointer, something else internally could have gone wrong.